File: sdk/py/src/agent_receipts/emitters/in_memory.py
"""InMemoryEmitter — array-backed test double (ADR-0020).

Performs no I/O and provides no delivery guarantee. Use in unit and
integration tests where the assertion is against the receipts that
passed through the emitter, not against a remote collector.

NOT for production use.
"""

from __future__ import annotations

from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from agent_receipts.receipt.types import AgentReceipt


class InMemoryEmitter:
    """Captures emitted receipts in an exposed list. Test double only."""

    def __init__(self) -> None:
        self._received: list[AgentReceipt] = []

    @property
    def received(self) -> list[AgentReceipt]:
        """All receipts passed to :meth:`emit`, in arrival order.

        Returned as a live list reference for cheap test assertions; copy
        before mutating if you need an independent snapshot.
        """
        return self._received

    def emit(self, receipt: AgentReceipt) -> None:
        self._received.append(receipt)

    def clear(self) -> None:
        """Drop all recorded receipts. Useful between test cases."""
        self._received.clear()

File: sdk/ts/src/emitters/in-memory.ts
/**
 * InMemoryEmitter — test double that captures emitted receipts in an
 * exposed array.
 *
 * Performs no I/O and provides no delivery guarantee. Use this in unit and
 * integration tests where the assertion is against the receipts that
 * passed through the emitter, not against a remote collector.
 *
 * NOT for production use.
 */

import type { AgentReceipt } from "../receipt/types.js";
import type { Emitter } from "./types.js";

export class InMemoryEmitter implements Emitter {
	private readonly _received: AgentReceipt[] = [];

	/** All receipts passed to {@link emit}, in arrival order. */
	get received(): readonly AgentReceipt[] {
		return this._received;
	}

	emit(receipt: AgentReceipt): Promise<void> {
		this._received.push(receipt);
		return Promise.resolve();
	}

	/** Clear the recorded receipts. Useful between test cases. */
	clear(): void {
		this._received.length = 0;
	}
}
